src/dbnode/persist/fs/msgpack/stream_with_digest.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"hash"
+	"hash/adler32"
+)
+
+var (
+	// errChecksumMismatch returned when the calculated checksum doesn't match the stored checksum
+	errChecksumMismatch = errors.New("calculated checksum doesn't match stored checksum")
+)
+
+var _ DecoderStream = &decoderStreamWithDigest{}
+
+// decoderStreamWithDigest calculates the digest as it processes a decoder stream.
+type decoderStreamWithDigest struct {
+	reader       DecoderStream
+	readerDigest hash.Hash32
+	unreadByte   bool
+	enabled      bool
+}
+
+// newDecoderStreamWithDigest returns a new decoderStreamWithDigest
+func newDecoderStreamWithDigest(reader DecoderStream) *decoderStreamWithDigest {
+	return &decoderStreamWithDigest{
+		reader:       reader,
+		readerDigest: adler32.New(),
+	}
+}
+
+func (d *decoderStreamWithDigest) Read(p []byte) (n int, err error) {
+	n, err = d.reader.Read(p)
+	if err != nil {
+		return n, err
+	}
+	if n <= 0 {
+		return n, nil
+	}
+
+	start := 0
+	if d.unreadByte {
+		d.unreadByte = false
+		start++
+	}
+	if d.enabled {
+		if _, err := d.readerDigest.Write(p[start:n]); err != nil {
+			return 0, err
+		}
+	}
+	return n, err
+}
+
+func (d *decoderStreamWithDigest) ReadByte() (byte, error) {
+	b, err := d.reader.ReadByte()
+	if err != nil {
+		return 0, err
+	}
+
+	if d.unreadByte {
+		d.unreadByte = false
+	} else if d.enabled {
+		if _, err := d.readerDigest.Write([]byte{b}); err != nil {
```

Comment:
I would add a single byte slice to the struct so it can be reused with an array allocation here:
```
type decoderStreamWithDigest struct {
	reader       DecoderStream
	readerDigest hash.Hash32
	unreadByte   bool
	enabled      bool
        singleByteBuff []byte
}

func newDecoderStreamWithDigest(...) {
  return &decoderStreamWithDigest{
    // ....
    singleByteBuff: make([]byte, 1),
  }
}


// Then to use it
d.singleByteBuff[0] = b
d.readerDigest.Write(d.singleByteBuff)
```

Reply:
Clever!